### PR TITLE
Wrap host property test and equivalent test function

### PR DESCRIPTION
**Pull request by qindazhu**

### CODE REVIEW

File: k2/csrc/host_shim.cu [hunk 1]
```diff
@@ -111,4 +110,52 @@ FsaVec FsaVecCreator::GetFsaVec() {
       arcs_);
 }
 
+Array1<bool> CheckProperties(FsaOrVec &fsas, bool (*f)(const k2host::Fsa &)) {
```
Comment (csukuangfj) [inline k2/csrc/host_shim.cu:113]: Can we rename to`GetFsaVecBasicProperties`? To be consistent with the one in k2/csrc/fsa.h

`Check` is not a good name here as the program will not abort.

File: k2/csrc/host_shim.h [hunk 2]
```diff
@@ -179,6 +182,163 @@ class FsaVecCreator {
...
+  function `f` which is one of property test functions for k2host::Fsa in
+  host/properties.h, e.g. IsValid(const k2::host Fsa&), IsTopSorted(const
+  k2host::Fsa&).
+
+  If `fsas` is FsaVec, the function will return an array on CPU which has
+    ans[i] = f(FsaVecToHostFsa(fsa_vec, i)) for 0 <= i < fsa_vec.Dim0();
+  else
+    returns a CPU array with size 1 and ans[0] = f(FsaToHostFsa(fsas))
+*/
+Array1<bool> CheckProperties(FsaOrVec &fsas, bool (*f)(const k2host::Fsa &));
```
Comment (csukuangfj) [inline k2/csrc/host_shim.h:204]: I would suggest moving it to host_shim.cc and giving it a local linkage.
Users will not invoke it directly.

File: k2/csrc/host_shim.h [hunk 3]
```diff
@@ -179,6 +182,163 @@ class FsaVecCreator {
...
+
+/*
+  ans[i]=true if `fsas[i]` is valid. An Fsa is valid if:
+  1. It is a valid Ragged array.
+  2. If it's not empty, it contains at least two states.
+     Noted empty fsa is valid.
+  3. Only arcs with symbol==-1 enter the final state.
+  4. There's no arcs leaving final_state.
```
Comment (csukuangfj) [inline k2/csrc/host_shim.h:212]: There're

File: k2/csrc/host_shim.h [hunk 4]
```diff
@@ -179,6 +182,163 @@ class FsaVecCreator {
...
+  5. All arcs leaving a state have a same src_state, which is the corresponding
+     idx0 in fsa, see index naming convention explained in utils.h.
```
Comment (csukuangfj) [inline k2/csrc/host_shim.h:214]: This can be idx01 if the input is a FsaVec.

File: k2/csrc/host_shim.h [hunk 5]
```diff
@@ -179,6 +182,163 @@ class FsaVecCreator {
...
+Array1<bool> CheckProperties(FsaOrVec &fsas, bool (*f)(const k2host::Fsa &));
+
+/*
+  ans[i]=true if `fsas[i]` is valid. An Fsa is valid if:
+  1. It is a valid Ragged array.
+  2. If it's not empty, it contains at least two states.
+     Noted empty fsa is valid.
+  3. Only arcs with symbol==-1 enter the final state.
+  4. There's no arcs leaving final_state.
+  5. All arcs leaving a state have a same src_state, which is the corresponding
```
Comment (csukuangfj) [inline k2/csrc/host_shim.h:213]: 6. The final state is numerically greater than any other state.

File: k2/csrc/host_shim.h [hunk 6]
```diff
@@ -179,6 +182,163 @@ class FsaVecCreator {
...
+/*
+  ans[i]=true if `fsas[i]` is acyclic.
+*/
+inline Array1<bool> IsAcyclic(FsaOrVec &fsas) {
+  return CheckProperties(fsas, IsAcyclicWapper);
+}
+
+/*
+  ans[i]=true if `fsas[i]` is deterministic; an Fsa is deterministic if
+  it has is no state that has multiple arcs leaving it with the same label on
```
Comment (csukuangfj) [inline k2/csrc/host_shim.h:262]: ~~is~~

File: k2/csrc/host_shim.h [hunk 7]
```diff
@@ -179,6 +182,163 @@ class FsaVecCreator {
...
+
+  @param [in]  a          One of the FSAs to be checked the equivalence.
+                          Must have NumAxes() == 2 and on CPU.
+  @param [in]  b          The other FSA to be checked the equivalence.
+                          Must have NumAxes() == 2 and on CPU.
+  @param [in]  top_sorted The user may set this to true if both `a` and `b` are
+                          topologically sorted; this makes this function faster.
+                          Otherwise it must be set to false.
+  @param [in]  log_semiring If true, the algorithm will only check paths
+                          within `beam` of the of the log-sum probs over
```
Comment (csukuangfj) [inline k2/csrc/host_shim.h:313]: of the ~~of~~

File: k2/csrc/host_shim.h [hunk 8]
```diff
@@ -179,6 +182,163 @@ class FsaVecCreator {
...
+                          will only check paths within `beam` of the
+                          best path(for tropical semiring, it's max
+                          weight over all paths from start state to
+                          final state; for log semiring, it's log-sum probs
+                          over all paths) in `a` or `b`. That is,
+                          any symbol sequence, whose total weights
+                          over all paths are within `beam` of the best
+                          path (either in `a` or `b`), must have
+                          the same weights in `a` and `b`.
+                          There is no any requirement on symbol sequences
```
Comment (csukuangfj) [inline k2/csrc/host_shim.h:328]: There is no ~~any~~ requirement

File: k2/csrc/host_shim.cu (hunk 1 above)
Comment (qindazhu) [reply k2/csrc/host_shim.cu:113]: Let's keep Check, as it just check one of those properties (GetFsaVecBasicProperties get all properties once)

File: k2/csrc/host_shim.h (hunk 2 above)
Comment (qindazhu) [reply k2/csrc/host_shim.h:204]: Sure, Will do

File: k2/csrc/host_shim.cu (hunk 1 above)
Comment (qindazhu) [reply k2/csrc/host_shim.cu:113]: Will move the function to .cu as a local linkage

File: k2/csrc/host_shim.h (hunk 4 above)
Comment (qindazhu) [reply k2/csrc/host_shim.h:214]: I explain it here starting with ` An Fsa is valid if`. Will add `idx1` (it's not idx01)

File: k2/csrc/host_shim.h (hunk 5 above)
Comment (qindazhu) [reply k2/csrc/host_shim.h:213]: Added

**Comment by qindazhu** [comment]
Ready to merge if there's no other issue.

**Comment by csukuangfj** [comment]
+2

Friday, 23 October 2020, 20:43 +0800 from notifications@github.com  <notifications@github.com>:

**Pull request merged**